inf-ruby.el

```diff
@@ -282,6 +284,20 @@ The following commands are available:
       (inf-ruby-remove-in-string (buffer-substring (point) end)
                                  inf-ruby-prompt-pattern))))
 
+(defun inf-ruby-buffer ()
+  "Return inf-ruby buffer for the current buffer or project."
+  (catch 'buffer
+    (let ((current-dir         (file-truename default-directory))
+          (current-project-dir (file-truename
+                                (locate-dominating-file default-directory
+                                                        #'inf-ruby-console-match))))
```

Comment:
This is a rather good idea, except people can also call a specific `inf-ruby-console-*` command to pick the kind of project they're in. I wonder if making the choice here can go against those.

Further, some users could also have their own commands that call `run-ruby`. If they just set `default-directory` before calling `run-ruby`, maybe that's the only way they need to communicate their project root.

On the other hand, we could go the other way and leave `inf-ruby-console-auto` as the only interactive command. Then going through `inf-ruby-console-patterns-alist` will be the obvious way to define new project types.